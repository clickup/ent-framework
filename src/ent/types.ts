import type { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  ID,
  LoadByInput,
  Order,
  Table,
  UniqueKey,
  Where,
} from "../types";
import type { Validation } from "./Validation";
import type { VC } from "./VC";

/**
 * A very shallow interface of Ent class (as a collection of static methods).
 * Used in some places where we need the very minimum from the Ent.
 */
export interface EntClass<TTable extends Table = any> {
  readonly SCHEMA: Schema<TTable>;
  readonly VALIDATION: Validation<TTable>;
  readonly name: string; // class constructor name

  new (): Ent;
  loadX(vc: VC, id: string): Promise<Ent>;
  loadNullable(vc: VC, id: string): Promise<Ent | null>;
  loadIfReadableNullable(vc: VC, id: string): Promise<Ent | null>;
  count(vc: VC, where: CountInput<TTable>): Promise<number>;
  exists(vc: VC, where: ExistsInput<TTable>): Promise<boolean>;
  select(
    vc: VC,
    where: Where<TTable>,
    limit: number,
    order?: Order<TTable>
  ): Promise<Ent[]>;
  selectChunked(
    vc: VC,
    where: Where<TTable>,
    chunkSize: number,
    limit: number,
    custom?: {}
  ): AsyncIterableIterator<Ent[]>;
  loadByX(vc: VC, keys: LoadByInput<TTable, UniqueKey<TTable>>): Promise<Ent>;
}

/**
 * A very shallow interface of one Ent.
 */
export interface Ent {
  readonly [ID]: string;
  readonly vc: VC;
  deleteOriginal(): Promise<boolean>;
}
