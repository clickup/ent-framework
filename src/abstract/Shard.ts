import type { Client } from "./Client";
import type { Query } from "./Query";
import type { QueryAnnotation, WhyClient } from "./QueryAnnotation";
import type { Timeline } from "./Timeline";

/**
 * Master freshness: reads always go to master.
 */
export const MASTER = Symbol("MASTER");

/**
 * Stale replica freshness: reads always go to a replica, even if it's stale.
 */
export const STALE_REPLICA = Symbol("STALE_REPLICA");

/**
 * Options passed to Shard constructor.
 */
export interface ShardOptions<TClient extends Client> {
  locateClient: (
    freshness: typeof MASTER | typeof STALE_REPLICA
  ) => Promise<TClient>;
  onRunError: (attempt: number, error: unknown) => Promise<"retry" | "throw">;
}

/**
 * Shard is a numbered island with one master and N replicas.
 */
export class Shard<TClient extends Client> {
  private shardClients = new WeakMap<TClient, TClient>();

  constructor(
    public readonly no: number,
    private options: ShardOptions<TClient>
  ) {}

  /**
   * Chooses the right client to be used for this shard. We don't memoize,
   * because the Shard may relocate to another Island during re-discovery.
   */
  async client(
    timeline: Timeline | typeof MASTER | typeof STALE_REPLICA
  ): Promise<TClient> {
    const { client } = await this.clientEx(timeline, undefined);
    return client;
  }

  /**
   * Runs a query after choosing the right client (destination connection,
   * shard, annotation etc.)
   */
  async run<TOutput>(
    query: Query<TOutput>,
    annotation: QueryAnnotation,
    timeline: Timeline,
    freshness: null | typeof MASTER | typeof STALE_REPLICA
  ): Promise<TOutput> {
    for (let attempt = 0; ; attempt++) {
      const { client, whyClient } = await this.clientEx(
        freshness ?? timeline,
        query.IS_WRITE ? true : undefined
      );

      let res;
      try {
        res = await query.run(client, {
          ...annotation,
          whyClient,
          attempt: annotation.attempt + attempt,
        });
      } catch (error: unknown) {
        // The shard was there by the moment we got its client, but it probably
        // disappeared (during migration) and appeared on some other island.
        if ((await this.options.onRunError(attempt, error)) === "retry") {
          continue;
        } else {
          throw error;
        }
      }

      if (query.IS_WRITE && freshness !== STALE_REPLICA) {
        timeline.setPos(
          await client.timelineManager.currentPos(),
          client.timelineManager.maxLagMs
        );
      }

      return res;
    }
  }

  /**
   * Throws if this shard does not exist, or its island is down, or something
   * else is wrong with it.
   */
  async assertDiscoverable(): Promise<void> {
    await this.options.locateClient(MASTER);
  }

  /**
   * An extended client selection logic. There are multiple reasons (8+ in total
   * so far) why a master or a replica may be chosen to send the query to. We
   * don't memoize, because the Shard may relocate to another Island during
   * re-discovery.
   */
  private async clientEx(
    timeline: Timeline | typeof MASTER | typeof STALE_REPLICA,
    isWrite: true | undefined
  ): Promise<{ client: TClient; whyClient: WhyClient }> {
    if (isWrite) {
      return {
        client: await this.shardClient(MASTER),
        whyClient: "master-bc-is-write",
      };
    }

    if (timeline === MASTER) {
      return {
        client: await this.shardClient(MASTER),
        whyClient: "master-bc-master-freshness",
      };
    }

    const replica = await this.shardClient(STALE_REPLICA);

    if (replica.isMaster) {
      return {
        client: replica,
        whyClient: "master-bc-no-replicas",
      };
    }

    if (timeline === STALE_REPLICA) {
      return {
        client: replica,
        whyClient: "replica-bc-stale-replica-freshness",
      };
    }

    const isCaughtUp = timeline.isCaughtUp(
      await replica.timelineManager.currentPos()
    );
    return isCaughtUp
      ? {
          client: replica,
          whyClient: isCaughtUp,
        }
      : {
          client: await this.shardClient(MASTER),
          whyClient: "master-bc-replica-not-caught-up",
        };
  }

  /**
   * Returns a shard-aware Client of a particular freshness.
   */
  private async shardClient(
    freshness: typeof MASTER | typeof STALE_REPLICA
  ): Promise<TClient> {
    const client = await this.options.locateClient(freshness);
    let shardClient = this.shardClients.get(client);
    if (!shardClient) {
      shardClient = client.withShard(this.no);
      this.shardClients.set(client, shardClient);
    }

    return shardClient;
  }
}
