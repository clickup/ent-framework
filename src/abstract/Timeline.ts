const SEP = ":";

/**
 * The reason why the decision that this replica timeline is "good enough" has
 * been made.
 */
export type TimelineCaughtUpReason =
  | false
  | "replica-bc-master-state-unknown"
  | "replica-bc-caught-up"
  | "replica-bc-pos-expired";

/**
 * Even when pos is expired, we still continue to serialize it for some time for
 * better debugging. Having this gap allows the system trigger
 * "replica-bc-pos-expired" reason of a replica choice longer.
 */
const SERIALIZE_EXPIRATION_GAP_MS = 600 * 1000;

/**
 * Tracks replication lag timeline position at master per "user" and Ent.
 * - serialization format: "pos:expiresAt"
 * - wipes expired records (expiration is calculated at assignment moment)
 *
 * How replication lag (timeline) tracking works: for each
 * microshard+Ent+"user", we know the “last write-ahead log write position”
 * which that user (typically, VC#principal) made recently. This info can be
 * propagated through e.g. user's session and push notifications/subscriptions
 * channels automatically (“serialized timeline” and “timelines merging”). So
 * the next time the same user tries to read the data from the same Ent on the
 * same microshard, Ent Framework makes a choice, whether the replica is “good
 * enough” for this already; if not, it falls back to master read. I.e. the data
 * is not granular to individual Ent ID, it’s granular to the
 * user+Ent+microshard, and thus it is decoupled from IDs.
 */
export class Timeline {
  constructor(
    private state:
      | "unknown"
      | { readonly pos: bigint; readonly expiresAt: number } = "unknown",
  ) {}

  static deserialize(
    data: string | undefined,
    prevTimeline: Timeline | null,
  ): Timeline {
    const parts = data ? data.split(SEP) : [];
    const pos = BigInt(parts[0] || "0");

    if (
      prevTimeline &&
      prevTimeline.state !== "unknown" &&
      prevTimeline.state.pos >= pos
    ) {
      // The previous timeline holds a more recent WAL position than the one
      // we're deserializing, so we should respect it better.
      return prevTimeline;
    }

    return new this(
      pos ? { pos, expiresAt: parseInt(parts[1] || "0") } : "unknown",
    );
  }

  static cloneMap(
    timelines: ReadonlyMap<string, Timeline>,
  ): Map<string, Timeline> {
    const copy = new Map<string, Timeline>();
    for (const [key, timeline] of timelines.entries()) {
      if (timeline.state !== "unknown") {
        copy.set(key, new Timeline(timeline.state));
      }
    }

    return copy;
  }

  serialize(): string | undefined {
    return this.state === "unknown" ||
      Date.now() >= this.state.expiresAt + SERIALIZE_EXPIRATION_GAP_MS
      ? undefined
      : this.state.pos.toString() + SEP + this.state.expiresAt;
  }

  setPos(pos: bigint, maxLagMs: number): void {
    if (this.state !== "unknown" && this.state.pos >= pos) {
      // We already hold a "more recent" pos, so don't need to update it.
      return;
    }

    this.state = { pos, expiresAt: Date.now() + maxLagMs };
  }

  isCaughtUp(replicaPos: bigint): TimelineCaughtUpReason {
    return this.state === "unknown"
      ? "replica-bc-master-state-unknown"
      : replicaPos >= this.state.pos
        ? "replica-bc-caught-up"
        : Date.now() >= this.state.expiresAt
          ? "replica-bc-pos-expired"
          : false;
  }

  reset(): void {
    this.state = "unknown";
  }
}
