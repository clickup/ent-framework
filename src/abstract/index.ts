/**
 * @file Automatically generated by barrelsby.
 */

export * from "./Batcher";
export * from "./Client";
export * from "./ClientError";
export * from "./Cluster";
export * from "./Island";
export * from "./Loader";
export * from "./LocalCache";
export * from "./Loggers";
export * from "./Query";
export * from "./QueryAnnotation";
export * from "./QueryBase";
export * from "./QueryPing";
export * from "./Runner";
export * from "./Schema";
export * from "./Shard";
export * from "./ShardError";
export * from "./ShardIsNotDiscoverableError";
export * from "./ShardNamer";
export * from "./Timeline";
export * from "./TimelineManager";
