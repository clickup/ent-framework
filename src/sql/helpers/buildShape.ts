const RE_COMMENT = toRegExp`
  /\*.*?\*/
`;

const RE_NUMBER = toRegExp`
  (?<!\w)\d+
`;

const RE_STRING = toRegExp`
  ' (?:[^']|'')* '
`;

const RE_IN_ROW = toRegExp`
  (
    \sIN\(
    ROW \( [^)]+ \)
  )
  (?: ,\s* ROW \( [^)]+ \) )*
`;

const RE_WITH_VALUES = toRegExp`
  ^(
    WITH \s [^\n]+VALUES\n
      [^\n]+,\n
      ( \s* \( (?:[^\n]+) \) )
  ),\n
    (?: \2 (?:,\n)? )+
`;

const RE_INSERT_VALUES = toRegExp`
  ^(
    INSERT \s INTO \s [^\n]+VALUES\n
      ( \s* \( (?:[^\n]+) \) )
  ),\n
    (?: \2 (?:,\n)? )+
`;

const RE_IDENTICAL_UNION_ALL = toRegExp`
  (?<=^|\n)
  ( \(? SELECT \s .*? )\n
  (?:
    (\s* UNION \s ALL)\n
    \1 \n?
  )+
`;

export default function buildShape(sql: string) {
  return sql
    .trim()
    .replace(RE_COMMENT, "")
    .replace(RE_NUMBER, "?")
    .replace(RE_STRING, "'?'")
    .replace(RE_IN_ROW, "$1, ...")
    .replace(RE_WITH_VALUES, "$1, ...")
    .replace(RE_INSERT_VALUES, "$1, ...")
    .replace(RE_IDENTICAL_UNION_ALL, "$1\n$2 ...");
}

function toRegExp(template: TemplateStringsArray) {
  return new RegExp(template.raw.join("").replace(/\s+/g, ""), "g");
}
