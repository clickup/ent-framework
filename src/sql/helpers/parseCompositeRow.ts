/**
 * Parses composite row input into parts. See references at:
 * - https://www.postgresql.org/docs/current/rowtypes.html
 * - https://github.com/DmitryKoterov/db_type/blob/master/lib/DB/Type/Pgsql/Row.php
 * - unit tests of this function
 */
export default function parseCompositeRow(str: string): Array<string | null> {
  let p = 0;
  let c = "";

  function readCharAfterSpaces(): string {
    while (p < str.length) {
      const c = str[p];
      if (c !== " " && c !== "\t" && c !== "\r" && c !== "\n") {
        break;
      }
    }

    return p < str.length ? str[p] : "";
  }

  function throwError(message: string) {
    throw Error(`${message} at position ${p} of "${str}"`);
  }

  // Leading "(".
  c = readCharAfterSpaces();
  if (c !== "(") {
    throwError('An anonymous composite type row must start with "("');
  }

  p++;

  // Check for immediate trailing ')': by convention, "()" translates to
  // ROW(NULL) tuple and not to an empty tuple (because ROW() and ROW(NULL) have
  // the same stringified representation which is "()" unfortunately).
  c = readCharAfterSpaces();
  if (c === ")") {
    return [null];
  }

  // Row may contain:
  // - "-quoted strings (escaping: ["] is doubled)
  // - unquoted strings (before first "," or ")")
  // - empty string (it is treated as NULL)
  // Nested rows and all other things are represented as strings.

  const reTillNext = /[^,)]*/sy;
  const reQuoted = /"((?:[^"]+|"")*)"/sy;

  const result: Array<string | null> = [];
  while (true) {
    // We read a value in this iteration, then - delimiter.
    c = readCharAfterSpaces();

    // Always read a next element value.
    if (c === "," || c === ")") {
      // Comma or end of row instead of value: treat as NULL.
      result.push(null);
    } else if (c !== '"') {
      // Unquoted string. Notice that NULL here is treated as "NULL" string, but
      // NOT as a null value! This is how composite values are encoded by PG.
      reTillNext.lastIndex = p;
      const matches = reTillNext.exec(str)!;
      result.push(matches[0]);
      p += matches[0].length;
    } else {
      reQuoted.lastIndex = p;
      const matches = reQuoted.exec(str);
      if (matches) {
        // Quoted string.
        result.push(matches[1].replace('""', '"').replace("\\\\", "\\"));
        p += matches[0].length;
      } else {
        throwError("Expected a balanced quoted string");
      }
    }

    // Delimiter or the end of row.
    c = readCharAfterSpaces();
    if (c === ",") {
      p++;
      continue;
    } else if (c === ")") {
      p++;
      break;
    } else {
      throwError('Expected a delimiter "," or ")"');
    }
  }

  return result;
}
