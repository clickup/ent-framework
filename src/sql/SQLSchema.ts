import type { Query } from "../abstract/Query";
import { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  InsertInput,
  LoadByInput,
  Row,
  SelectByInput,
  SelectInput,
  Table,
  UniqueKey,
  UpdateInput,
} from "../types";
import { SQLQueryCount } from "./SQLQueryCount";
import { SQLQueryDelete } from "./SQLQueryDelete";
import { SQLQueryExists } from "./SQLQueryExists";
import { SQLQueryIDGen } from "./SQLQueryIDGen";
import { SQLQueryInsert } from "./SQLQueryInsert";
import { SQLQueryLoad } from "./SQLQueryLoad";
import { SQLQueryLoadBy } from "./SQLQueryLoadBy";
import { SQLQuerySelect } from "./SQLQuerySelect";
import { SQLQuerySelectBy } from "./SQLQuerySelectBy";
import { SQLQueryUpdate } from "./SQLQueryUpdate";
import { SQLQueryUpsert } from "./SQLQueryUpsert";

export class SQLSchema<
  TTable extends Table,
  TUniqueKey extends UniqueKey<TTable>
> extends Schema<TTable, TUniqueKey> {
  idGen(): Query<string> {
    return new SQLQueryIDGen(this);
  }

  insert(input: InsertInput<TTable>): Query<string | null> {
    return new SQLQueryInsert(this, input);
  }

  upsert(input: InsertInput<TTable>): Query<string> {
    return new SQLQueryUpsert(this, input);
  }

  update(id: string, input: UpdateInput<TTable>): Query<boolean> {
    return new SQLQueryUpdate(this, id, input);
  }

  delete(id: string): Query<boolean> {
    return new SQLQueryDelete(this, id);
  }

  load(id: string): Query<Row<TTable> | null> {
    return new SQLQueryLoad(this, id);
  }

  loadBy(input: LoadByInput<TTable, TUniqueKey>): Query<Row<TTable> | null> {
    return new SQLQueryLoadBy(this, input);
  }

  selectBy(
    input: SelectByInput<TTable, TUniqueKey>
  ): Query<Array<Row<TTable>>> {
    return new SQLQuerySelectBy(this, input);
  }

  select(input: SelectInput<TTable>): Query<Array<Row<TTable>>> {
    return new SQLQuerySelect(this, input);
  }

  count(input: CountInput<TTable>): Query<number> {
    return new SQLQueryCount(this, input);
  }

  exists(input: ExistsInput<TTable>): Query<boolean> {
    return new SQLQueryExists(this, input);
  }
}
