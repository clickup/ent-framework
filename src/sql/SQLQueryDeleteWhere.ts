import { QueryBase } from "../abstract/Query";
import { QueryAnnotation } from "../abstract/QueryAnnotation";
import { Schema } from "../abstract/Schema";
import { hash } from "../helpers";
import { DeleteWhereInput, ID, Table } from "../types";
import { SQLClient } from "./SQLClient";
import { SQLRunner } from "./SQLRunner";

export class SQLQueryDeleteWhere<TTable extends Table> extends QueryBase<
  TTable,
  DeleteWhereInput<TTable>,
  string[],
  SQLClient
> {
  protected readonly RUNNER_CLASS = SQLRunnerDeleteWhere;
}

export class SQLRunnerDeleteWhere<TTable extends Table> extends SQLRunner<
  TTable,
  DeleteWhereInput<TTable>,
  string[]
> {
  static override readonly IS_WRITE = true;
  readonly op = "DELETE_WHERE";
  readonly default = [];

  private builder;

  constructor(schema: Schema<TTable>, client: SQLClient) {
    super(schema, client);
    this.builder = {
      prefix: this.fmt("DELETE FROM %T"),
      func: (input: DeleteWhereInput<TTable>) =>
        this.buildOptionalWhere(this.schema.table, input),
      suffix: this.fmt(` RETURNING %PK AS ${ID}`),
    };
  }

  override key(input: DeleteWhereInput<TTable>): string {
    // Coalesce equal delete queries.
    const json = JSON.stringify(input);
    return hash(json);
  }

  async runSingle(
    input: DeleteWhereInput<TTable>,
    annotations: QueryAnnotation[]
  ): Promise<string[]> {
    if (!(input[ID] instanceof Array)) {
      throw Error(
        `Field ${ID} must be an array of IDs in ${this.op} query (for safety)`
      );
    }

    const sql =
      this.builder.prefix + this.builder.func(input) + this.builder.suffix;
    const rows = await this.clientQuery<{ [ID]: string }>(
      sql,
      annotations,
      input[ID].length
    );
    return rows.map((row) => row.id);
  }

  runBatch = undefined;
}
