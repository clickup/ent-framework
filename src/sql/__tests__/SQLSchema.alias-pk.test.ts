import type { Query } from "../../abstract/Query";
import type { Shard } from "../../abstract/Shard";
import { MASTER } from "../../abstract/Shard";
import { Timeline } from "../../abstract/Timeline";
import { nullthrows } from "../../helpers";
import { ID } from "../../types";
import { SQLQueryDeleteWhere } from "../SQLQueryDeleteWhere";
import { SQLSchema } from "../SQLSchema";
import type { TestSQLClient } from "./helpers/TestSQLClient";
import { testCluster } from "./helpers/TestSQLClient";

const TABLE = 'schema"test_alias';
const timeline = new Timeline();
let shard: Shard<TestSQLClient>;
let master: TestSQLClient;

beforeEach(async () => {
  timeline.reset();
  shard = testCluster.randomShard();
  master = await shard.client(MASTER);
  await master.rows("DROP TABLE IF EXISTS %T CASCADE", TABLE);
  await master.rows(
    `CREATE TABLE %T(
      user_id bigint NOT NULL PRIMARY KEY,
      name text NOT NULL
    )`,
    TABLE
  );
});

const schema = new SQLSchema(
  TABLE,
  {
    user_id: { type: ID, autoInsert: "id_gen()" },
    name: { type: String },
  },
  ["user_id"]
);

async function shardRun<TOutput>(query: Query<TOutput>) {
  return shard.run(
    query,
    {
      trace: "some-trace",
      rawTrace: "123456434",
      debugStack: "",
      vc: "some-vc",
      whyClient: undefined,
    },
    timeline,
    null
  );
}

test("ops_single", async () => {
  const id1 = await shardRun(schema.insert({ user_id: "1", name: "n1" }));
  master.toMatchSnapshot();
  expect(id1).toEqual("1");

  const id2 = await shardRun(schema.upsert({ user_id: "2", name: "n2" }));
  master.toMatchSnapshot();

  const id3 = await shardRun(schema.insert({ name: "n3" }));
  master.toMatchSnapshot();

  {
    const res = await shardRun(schema.update(nullthrows(id1), {})); // no DB query is sent here
    master.toMatchSnapshot();
    expect(res).toEqual(true);
  }

  {
    const res = await shardRun(schema.upsert({ user_id: "101", name: "n11" }));
    master.toMatchSnapshot();
    expect(res).toEqual("101");
  }

  {
    const res = await shardRun(schema.delete(nullthrows(id1)));
    master.toMatchSnapshot();
    expect(res).toEqual(true);
  }

  {
    const res = await shardRun(
      new SQLQueryDeleteWhere(schema, { id: [id2!], $literal: ["1=1"] })
    );
    master.toMatchSnapshot();
    expect(res).toEqual([id2]);
  }

  {
    const res = await shardRun(schema.load(id3!));
    master.toMatchSnapshot();
    expect(res).toMatchObject({ id: id3, name: "n3" });
  }

  {
    const res = await shardRun(schema.loadBy({ user_id: id3! }));
    master.toMatchSnapshot();
    expect(res).toMatchObject({ id: id3, name: "n3" });
  }

  {
    const res = await shardRun(schema.count({ user_id: id3! }));
    master.toMatchSnapshot();
    expect(res).toEqual(1);
  }

  {
    const res = await shardRun(
      schema.select({ where: { user_id: id3! }, limit: 10 })
    );
    master.toMatchSnapshot();
    expect(res).toMatchObject([{ user_id: id3 }]);
  }
});
