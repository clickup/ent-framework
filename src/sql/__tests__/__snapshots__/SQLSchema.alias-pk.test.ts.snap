// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single ops 1`] = `
"
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  ('1', 'n1') ON CONFLICT DO NOTHING RETURNING user_id AS id
----
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  ('?', '?') ON CONFLICT DO NOTHING RETURNING user_id AS id
"
`;

exports[`single ops 2`] = `
"
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  ('2', 'n2')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
----
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  ('?', '?')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
"
`;

exports[`single ops 3`] = `
"
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  (id_gen(), 'n3') ON CONFLICT DO NOTHING RETURNING user_id AS id
----
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  (id_gen(), '?') ON CONFLICT DO NOTHING RETURNING user_id AS id
"
`;

exports[`single ops 4`] = `""`;

exports[`single ops 5`] = `
"
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  ('101', 'n11')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
----
INSERT INTO "schema""test_alias" (user_id, name) VALUES
  ('?', '?')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
"
`;

exports[`single ops 6`] = `
"
DELETE FROM "schema""test_alias" WHERE user_id IN('1') RETURNING user_id AS id
----
DELETE FROM "schema""test_alias" WHERE user_id IN('?') RETURNING user_id AS id
"
`;

exports[`single ops 7`] = `
"
DELETE FROM "schema""test_alias" WHERE user_id IN('2') AND (1=1) RETURNING user_id AS id
----
DELETE FROM "schema""test_alias" WHERE user_id IN('?') AND (?=?) RETURNING user_id AS id
"
`;

exports[`single ops 8`] = `
"
SELECT user_id, name, user_id AS id FROM "schema""test_alias"
  WHERE user_id IN('<id>')
----
SELECT user_id, name, user_id AS id FROM "schema""test_alias"
  WHERE user_id IN('?')
"
`;

exports[`single ops 9`] = `
"
SELECT user_id, name, user_id AS id FROM "schema""test_alias"
  WHERE user_id='<id>'
----
SELECT user_id, name, user_id AS id FROM "schema""test_alias"
  WHERE user_id='?'
"
`;

exports[`single ops 10`] = `
"
SELECT COUNT(1) AS count FROM "schema""test_alias"
  WHERE user_id='<id>'
----
SELECT COUNT(?) AS count FROM "schema""test_alias"
  WHERE user_id='?'
"
`;

exports[`single ops 11`] = `
"
SELECT EXISTS (SELECT true FROM "schema""test_alias"
  WHERE user_id='<id>')
----
SELECT EXISTS (SELECT true FROM "schema""test_alias"
  WHERE user_id='?')
"
`;

exports[`single ops 12`] = `
"
SELECT user_id, name, user_id AS id FROM "schema""test_alias"
  WHERE user_id='<id>'
  LIMIT 10
----
SELECT user_id, name, user_id AS id FROM "schema""test_alias"
  WHERE user_id='?'
  LIMIT ?
"
`;
